include/onnxruntime/core/session/onnxruntime_cxx_inline.h

```diff
@@ -891,6 +891,39 @@ inline std::string CustomOpApi::KernelInfoGetAttribute<std::string>(_In_ const O
   return out;
 }
 
+template <>
+inline std::vector<float> CustomOpApi::KernelInfoGetAttributeArray(_In_ const OrtKernelInfo* info, _In_ const char* name) {
+  size_t size = 0;
+  std::vector<float> out;
+  OrtStatus* status = api_.KernelInfoGetAttributeArray_float(info, name, nullptr, &size);
+
+  // The status should be ORT_INVALID_ARGUMENT because the size is insufficient to hold the attribute array
+  if (api_.GetErrorCode(status) == ORT_INVALID_ARGUMENT) {
```

Comment:
>ORT_INVALID_ARGUMENT [](start = 35, length = 20)

This error code is too generic for the buffer size. If we are to adopt this semantics, then potentially other APIs would use it too and I vote for a separate error code for the condition

Reply:
Currently, we use the same logic for parsing string attributes and quite honestly it makes the C++ interface very simple for a user. In the C++ wrapper, we actually call the C API with size 0 and by definition, we will get back an ORT_INVALID_ARGUMENT and we use the returned "right size" to resize the vector before making the call again. 

As for other APIs potentially using them - we go a case by cases basis don't we ? I don't quite get the concern here. Can you please provide an illustrative scenario ?

Reply:
I think the logic is fine. Not sure about case by case, since this is already the second time are using this approach. Some APIs create a separate error code for the condition since any argument can be invalid. Let's say an API says INVALID argument bc name is nullptr. It returns the error code but instead of returning we attempt to allocate 0 len buffer. The next call will return again invalid argument. This does not sound good to me.

---
In reply to: [587093854](https://github.com/microsoft/onnxruntime/pull/6878#discussion_r587093854) [](ancestors = 587093854)

Reply:
The C API that this C++ API taps into returns an ORT_FAIL for cases the attribute is invalid. That being said, I understand your concern about "true" ORT_INVALID error code from the lower level C API and the "expected" ORT_INVALID error code that we want to see just to figure out what the true size of the underlying attribute is. So, I ll make it safer by branching into the logic that does the resizing only if the error code says ORT_INVALID and the error message contains what we return back from the C API to indicate that we need a larger buffer to hold the contents of the attribute.